lib/pecorino/leaky_bucket.rb

```diff
@@ -86,42 +86,17 @@ def initialize(key:, leak_rate:, capacity:)
   # @param n_tokens[Float]
   # @return [State] the state of the bucket after the operation
   def fillup(n_tokens)
-    add_tokens(n_tokens.to_f)
+    capped_level_after_fillup, did_overflow = Pecorino.adapter.add_tokens(capa: @capacity, key: @key, leak_rate: @leak_rate, n_tokens: n_tokens)
+    State.new(capped_level_after_fillup, did_overflow)
   end
 
   # Returns the current state of the bucket, containing the level and whether the bucket is full.
   # Calling this method will not perform any database writes.
   #
   # @return [State] the snapshotted state of the bucket at time of query
   def state
-    conn = ActiveRecord::Base.connection
-    query_params = {
-      key: @key,
-      capa: @capacity.to_f,
-      leak_rate: @leak_rate.to_f
-    }
-    # The `level` of the bucket is what got stored at `last_touched_at` time, and we can
-    # extrapolate from it to see how many tokens have leaked out since `last_touched_at` -
-    # we don't need to UPDATE the value in the bucket here
-    sql = ActiveRecord::Base.sanitize_sql_array([<<~SQL, query_params])
-      SELECT
-        GREATEST(
-          0.0, LEAST(
-            :capa,
-            t.level - (EXTRACT(EPOCH FROM (clock_timestamp() - t.last_touched_at)) * :leak_rate)
-          )
-        )
-      FROM 
-        pecorino_leaky_buckets AS t
-      WHERE
-        key = :key
-    SQL
-
-    # If the return value of the query is a NULL it means no such bucket exists,
-    # so we assume the bucket is empty
-    current_level = conn.uncached { conn.select_value(sql) } || 0.0
-
-    State.new(current_level, (@capacity - current_level).abs < 0.01)
+    current_level, is_full = Pecorino.adapter.state(key: @key, capa: @capacity, leak_rate: @leak_rate)
```

Comment:
**nitpick:** this is me being very picky, but it seems that "cap" is a better abbreviation to "capacity" than "capa". It's more commonly used in my experience.

I was at first slightly confused and had to double-check that "capa" is an abbreviation for “capacity”.